src/Workspaces/Core/Portable/Workspace/Solution/ProjectState.cs

```diff
@@ -821,20 +833,29 @@ public ProjectState RemoveDocuments(ImmutableArray<DocumentId> documentIds)
 
     public ProjectState RemoveAdditionalDocuments(ImmutableArray<DocumentId> documentIds)
     {
+        if (documentIds.IsEmpty)
+            return this;
+
         return With(
             projectInfo: ProjectInfo.WithVersion(Version.GetNewerVersion()),
             additionalDocumentStates: AdditionalDocumentStates.RemoveRange(documentIds));
     }
 
     public ProjectState RemoveAnalyzerConfigDocuments(ImmutableArray<DocumentId> documentIds)
     {
+        if (documentIds.IsEmpty)
+            return this;
+
         var newAnalyzerConfigDocumentStates = AnalyzerConfigDocumentStates.RemoveRange(documentIds);
 
         return CreateNewStateForChangedAnalyzerConfigDocuments(newAnalyzerConfigDocumentStates);
     }
 
     public ProjectState RemoveAllDocuments()
     {
+        if (DocumentStates.IsEmpty && AdditionalDocumentStates.IsEmpty && AnalyzerConfigDocumentStates.IsEmpty)
```

Comment:
i don't see why we need to do the IsEmpty checks any anything but DocumentStates. t he semantics of this method (based on its body) seems to be that it's just removing DocumentStates, nothing else.  so even if htere are additional dcs, we can short circuit.